src/media-utils.ts

```diff
@@ -233,35 +236,94 @@ export async function getEnumeratedDevices (sdk: PureCloudWebrtcSdk): Promise<IE
     sdk.logger.debug('Returning cached enumerated devices', { devices: enumeratedDevices });
     return enumeratedDevices;
   }
 
   try {
-    enumeratedDevices.videoDeviceIds = [];
-    enumeratedDevices.audioDeviceIds = [];
-    enumeratedDevices.outputDeviceIds = [];
+    const oldDevices = enumeratedDevices;
+    enumeratedDevices.videoDevices = [];
+    enumeratedDevices.audioDevices = [];
+    enumeratedDevices.outputDevices = [];
 
-    const devices = await window.navigator.mediaDevices.enumerateDevices();
+    const newDevices = await window.navigator.mediaDevices.enumerateDevices();
     refreshDevices = false;
-    devices.forEach(device => {
+
+    const mappedDevices = mapOldToNewDevices(oldDevices, newDevices);
+    mappedDevices.forEach(device => {
       if (device.kind === 'videoinput') {
-        enumeratedDevices.videoDeviceIds.push(device.deviceId);
+        enumeratedDevices.videoDevices.push(device);
       } else if (device.kind === 'audioinput') {
-        enumeratedDevices.audioDeviceIds.push(device.deviceId);
+        enumeratedDevices.audioDevices.push(device);
       } else /* if (device.kind === 'audiooutput') */ {
-        enumeratedDevices.outputDeviceIds.push(device.deviceId);
+        enumeratedDevices.outputDevices.push(device);
       }
     });
   } catch (e) {
     throwSdkError.call(sdk, SdkErrorTypes.generic, 'Error enumerating devices', e);
   }
 
+  sdk.logger.debug('Enumerated devices', { devices: enumeratedDevices });
   return enumeratedDevices;
 }
 
+function mapOldToNewDevices (oldEnumeratedDevices: IEnumeratedDevices | undefined, newDevices: MediaDeviceInfo[]): MediaDeviceInfo[] {
+  /* if we have labels on the new devices, no reason to do anything with them
+    OR if we didn't have old labels, no need to do anything */
+  if (hasDevicePermissions(newDevices) || !oldEnumeratedDevices || !hasDevicePermissions(oldEnumeratedDevices)) {
+    return newDevices;
+  }
+
+  /* if we don't have labels, but we HAD labels - we can map the old labels to the new devices without labels (because FF) */
+  const devices = [];
+  const oldDevices = mapSdkEnumeratedDevicesToArray(oldEnumeratedDevices);
+  for (const newDevice of newDevices) {
+    const foundOldDevice = oldDevices.find(d => d.deviceId === newDevice.deviceId && d.groupId === newDevice.groupId);
+    if (foundOldDevice && foundOldDevice.label) {
+      devices.push(oldDevices);
+    } else {
+      devices.push(newDevice);
+    }
+  }
+
+  return devices;
+}
+
+function mapSdkEnumeratedDevicesToArray (devices: IEnumeratedDevices): MediaDeviceInfo[] {
+  const arr = [];
+  const addDevices = (device: MediaDeviceInfo) => arr.push(device);
+
+  devices.audioDevices.forEach(addDevices);
+  devices.videoDevices.forEach(addDevices);
+  devices.outputDevices.forEach(addDevices);
+
+  return arr;
+}
+
+function isSdkEnumeratedDevices (obj: IEnumeratedDevices | MediaDeviceInfo[]): obj is IEnumeratedDevices {
```

Comment:
This function is just checking the type of the object. In `hasDevicePermissions()`, either `IEnumeratedDevices` _or_ `MediaDeviceInfo[]` can be passed in (depending on who calls it). This just makes sure we can have an array to work with. Mainly because we store the enumerated devices based on type and not as a single array.

Reply:
I may be misunderstanding your question, though. Let me know if I am and I can try to change my answer.

Reply:
There are two reasons for this: 
1. To make TypeScript happy. This is a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) so it will return the correct typing of the object. 
2. After the ts compiles out, we still have valid js to check if the passed object is an `IEnumeratedDevices` object or an array. 

It felt like a little bit of overkill at first, but after realizing that some functions should be able to work with the `IEnumeratedDevices` object _or_ the array of devices, I decided to keep it.

Reply:
so what's with this `obj is IEnumeratedDevices` instead of `boolean`?

Reply:
That is just a fancy way to tell TS what the result is. So it translates to this: 
If this function returns `true`, we have a `IEnumeratedDevices` type. If it returns `false`, we don't have that. 
It will still return a boolean. That is just some special syntax.